### PR TITLE
Add Payload::Clone and fix NewToOldSubscriber which was cloning unsafely

### PR DESCRIPTION
**Pull request by nickgg**
Added unit test for the new method - I have tested NewOldBridge manually.

### CODE REVIEW

File: src/Payload.cpp [hunk 1]
```diff
@@ -55,6 +55,18 @@ void Payload::clear() {
   metadata.reset();
 }
 
+Payload Payload::clone() const {
+	Payload out;
```
Comment (alexmalyshev) [inline src/Payload.cpp]: We use 2-space indents, I think these are tabs?

File: src/Payload.h [hunk 2]
```diff
@@ -34,6 +34,9 @@ struct Payload {
   std::string cloneDataToString() const;
   void clear();
 
+  // Clone this object onto the stack
```
Comment (alexmalyshev) [inline src/Payload.h]: Not sure the `onto the stack` part is helpful.  A Payload is a tuple of 2 heap pointers, might be confusing.  No good ideas on what to say instead though...

File: test/PayloadTest.cpp [hunk 3]
```diff
@@ -34,3 +34,47 @@ TEST(PayloadTest, GiantMetadata) {
       FrameSerializerV0_1::deserializeMetadataFrom(cur, FrameFlags::METADATA),
       std::runtime_error);
 }
+
+TEST(PayloadTest, Clone) {
```
Comment (alexmalyshev) [inline test/PayloadTest.cpp]: (y)

File: experimental/rsocket/OldNewBridge.h [hunk 4]
```diff
@@ -121,8 +121,7 @@ class NewToOldSubscriber : public yarpl::flowable::Subscriber<reactivesocket::Pa
   void onNext(const reactivesocket::Payload& payload) override {
     // Cloning IOBufs just shares their internal buffers, so this isn't the end
     // of the world.
-    reactivesocket::Payload clone(
-        payload.data->clone(), payload.metadata->clone());
+    reactivesocket::Payload clone = payload.clone();
```
Comment (alexmalyshev) [inline experimental/rsocket/OldNewBridge.h]: `auto clone = payload.clone();`

File: src/Payload.cpp [hunk 5]
```diff
@@ -55,6 +55,18 @@ void Payload::clear() {
...
+	if (data) {
+		out.data = data->clone();
```
Comment (lexs) [inline src/Payload.cpp]: tabs? 👎

File: src/Payload.cpp (hunk 1 above)
Comment (lexs) [inline src/Payload.cpp]:
```cpp
return Payload(
  data ? data->clone() : nullptr,
  metadata ? metadata->clone() : nullptr)
```
is a more concise and clearer version

File: src/Payload.cpp (hunk 5 above)
Comment (nickgg) [inline src/Payload.cpp]: Ah, I copied this out of phabricator - not sure why it came out like this.

File: src/Payload.cpp (hunk 1 above)
Comment (nickgg) [inline src/Payload.cpp]: I don't agree with clearer, you've packed two ternary operators into a variable intialisation.

Comment (lexs) [inline src/Payload.cpp]: To each his own I suppose. I felt it was clearer as we're no longer mutating the Payload. I don't feel strongly about it however and it'll compile to exactly the same code so ¯\_(ツ)_/¯

File: experimental/rsocket/OldNewBridge.h [hunk 6]
```diff
@@ -121,8 +121,7 @@ class NewToOldSubscriber : public yarpl::flowable::Subscriber<reactivesocket::Pa
   void onNext(const reactivesocket::Payload& payload) override {
     // Cloning IOBufs just shares their internal buffers, so this isn't the end
     // of the world.
-    reactivesocket::Payload clone(
-        payload.data->clone(), payload.metadata->clone());
+    auto clone = payload.clone();
     inner_->onNext(std::move(clone));
```
Comment (lexs) [inline experimental/rsocket/OldNewBridge.h]: nit: `inner_->onNext(payload.clone());`

**Pull request merged**